BLU: Extensive support

Hugmeir opened the pull request:
After making https://github.com/xivanalysis/xivanalysis/pull/1794 I ended up working on the BLU codebase a chunk more, so this MR is an evolution from that one.  The first MR is already big enough however, so I'm leaving it open in case y'all would prefer something more digestible than this monster.

This includes all of the changes from the original:
- Adds the Shadowbringers spells and most of the effects
- Marks `Waning Nocturne` as an unable-to-act action, fixing the GCD uptime report
- Tweaks the uptime report to understand that Phantom Flurry is a channel ala Flamethrower
- DoT uptime
- Cold Fog suggestions
- Moon Flute window panel
- Revenge Blast window panel
- Adds the standard Swiftcast/Interrupted cast/big cooldown/tincture reports

Plus the linter fixes from https://github.com/xivanalysis/xivanalysis/pull/1796

Plus several new goodies:
- The Surpanakha x4 weaving now counts as uptime
- Improved the weaving detection of Surpanakhas (in particular, penta or hexa weaving Surpanakhas during Moon Flutes no longer counts as a weaving issue)
- Dropped Surpanakha combos (not doing it 4 times, or mistakenly splitting the combo and thus dropping the buff) are now reported as a suggestion
- It nows reports on dropped Phantom Flurry ticks from cancelling the channel too early, or if using Phantom Flurry outside of a Moon Flute, it reports if the big kick was not pressed
- Both the Cooldown and Moon Flute reports now accept some alternative spells (J Kick / Quasar, Feather Rain / Eruption, Shock Strike / Mountain Buster)
- The core `Action` interface now has an (optional) `DamageType` attribute; all the damaging BLU spells have been marked as Physical/Magical/Dark
- BLU spells have also been marked with ASTRAL or UMBRAL
- Both of the above are now used for a report for each Buff; not just Off-guard and Peculiar Light, but also the three Condensed Libra buffs.
- Enabled the Overhealing report, though it is conditional on using non-White Wind healing abilities
- We now detect if `Triple Trident` was used without first buffing it with `Whistle` and `Tingle` and show a suggestion about it
- The `Condensed Libra` raid buffs now show on the timeline
- All the newly added translations tags are now in `locale/en/messages.json`
- I added myself as a contributor :)

Rather than taking pictures of each bit individually, attached are some full reports:
![Screenshot 2023-05-08 at 22-24-27 xivanalysis](https://user-images.githubusercontent.com/215227/236926783-fe4bf4e6-f7a6-4773-9b73-9f506295f643.png)

![Screenshot 2023-05-08 at 22-22-55 xivanalysis](https://user-images.githubusercontent.com/215227/236926534-e8238566-7743-44d4-a8a2-8c668badc294.png)

I think this covers just about everything BLU needs at the moment. Just in time for 6.45 to bitrot it all! :D

Hugmeir committed:
BLU: Mark the Waning Nocturne window as unable to act

For reasons only SE will ever understand, Waning Nocturne is not
marked with `LockActions: true`, even though it does very much
lock your actions.

This commit just manually adds the effect to the generator, and
then runs the script to update unableToActStatusIds.ts.

Hugmeir committed:
BLU: Update ACTIONS and STATUSES to include (most) things added back in 5.45

This is also making the (100% debatable judgement call) of tagging
Phantom Flurry as a GCD rather than an oGCD.

I'll freely admit that this is not accurate to reality, but it
makes the reports easier to understand -- There's always going
to be a gap with no actions after pressing the button, and this
makes the button show up in the GCD timeline.

Hugmeir committed:
BLU: ABC report should undersyand Phantom Flurry

Phantom Flurry is essentially MCH's Flamethrower, except that it
starts with an oGCD rather than a GCD.

This commit extends BLU's ABC report with essentially a copy of
MCH's ABC report, with the Flamethrowers replaced with Phantom Flurries.

Hugmeir committed:
BLU: Bleeding DoT uptime report

The Bleeding DoT (from Song of Torment and Nightbloom) is a significant
contributor to BLU's damage, so this adds a module to track it.

It also adds a (minor severity) suggestion to cast Bristle before
each Song of Torment refresh, as that's a minor potency gain.

Hugmeir committed:
BLU: Enable the standard Tincture report

Hugmeir committed:
BLU: Track several big cooldowns

Namely, we track these:

    Nightbloom
    Phantom Flurry
    Feather Rain
    Shock Strike
    Glass Dance
    J Kick
    Triple Trident
    The Rose of Destruction
    Cold Fog
    Off-guard/Peculiar Light

Plus two defensives:

    Addle
    Magic Hammer

Tracking Magic Hammer might be a bit unfair since not everyone
may be carrying it, but hey, it's a big defensive cooldown, if you
have it slotted you should be using it.

Hugmeir committed:
BLU: Add the Interrupted Casts report

Hugmeir committed:
BLU: Enable the standard Swiftcast report

Hugmeir committed:
BLU: Add suggestions for Cold Fog

Cold Fog works similarly to TBN or Third Eye -- you cast it and then
you need to take damage to proc it.

This adds three suggestions:

    Severity MAJOR: Letting Cold Fog expire without getting hit
    Severity MAJOR: Casting White Death only once, making Cold Fog
                    not worth it
    Severity MEDIUM: Casting White Death less than 6 times.

Hugmeir committed:
BLU: Added a Moon Flute report

This shows the actions taken after casting Moon Flute, and has
several columns with suggested actions that should likely be in
the window (e.g. Nightbloom, Triple Trident).

Hugmeir committed:
BLU: Revenge Blast report

Revenge Blast deals 500 potency when under 20% of our max HP, so
finding and optimizing our Revenge Blast windows is always good.

This report does some funny heuristics to find potential blast windows
(5+ seconds at >20% HP) to display in a nice report, as well as
tracking bad Revenge Blasts (cast when HP has above 20%) and adding
a minor optimization suggestion to Whistle before the first Revenge
Blast of each window.

It's still kind of finicky, and it doesn't handle downtime too well,
but as-is it already helped my group find a window we didn't know
existed.

Hugmeir committed:
BLU: minor linting fix

Hugmeir committed:
BLU: Improvements to how we handle weaving & Surpanakhas

Surpanakha is an oGCD with 4 charges.  Each time you use it,
it gives you a stacking buff that makes the next Surpanakha stronger,
but only IF the next action is a Surpanakha.

Doing absolutely anything other than a Surpanakha while the buff
is up will completely drop it.

Weaving before or after the Surpanakha spam will always result in
fully clipping the following GCD.

So far, this is what the existing Weaving module for BLU did; it
checked if you had a Surpanakha x4 weave and gave you a pass.

However, a problem: the standard BLU opener actually subverts this!
It hexaweaves the Surpanakhas to make them fit inside of a Moon Flute.

So to avoid some useless noise, these changes allows for a
penta/hexaweave during the Moon Flute window IF the Surpanakhas are
at the end of the weave.

And since I was already tweaking this, it also adds new suggestion
if Surpanakha is used in sets of less than 4, or if other abilities
are used in-between the sequence, dropping the buff.

Hugmeir committed:
BLU: add the missing action for the kick ability during Phantom Flurry

When Phantom Flurry is being channeled, it changes into a different
ability, also confusingly named Phantom Flurry (but with a different
icon).

Using Phantom Flurry will consume Phantom Flurry and cancel Phantom
Flurry.  Very clear, as we can all agree.

...but for those in the audience that didn't quite follow that madness:

Pressing Phantom Flurry (action 23288) gives the buff
Phantom Flurry (status 2502).

The buff will get dropped if any action or movement is taken.

While the buff is active, two things will happen.

First, it starts a channel ability that deals a 200 potency
attack after every second.

Second, it changes the button for Phantom Flurry (action 23288)
into a Phantom Flurry (action 23289), which if pressed,
deals a 600 potency attack and cancels the Phantom Flurry (status 2502) buff.

Hugmeir committed:
BLU: Track lost Phantom Flurry ticks

Phantom Flurry is a channel that deals a 200 potency hit for every
second that it is channeled.

DPSes generally want to channel the entire 5 second duration at the
end of the Moon Flute window, so losing even one or two ticks
is a fair bit of potency.

Non-DPSes on the other hand want to get as close as possible to 5
seconds, then press the button again for an extra 600 potency hit.

This is adding suggestions for both; one for missing ticks, and
one for not pressing the button again if outside of a moon flute.

Hugmeir committed:
BLU: Add Quasar and Eruption as alternatives for J Kick and Feather Rain

Shared cooldowns, identical or near-identical damage.

Hugmeir committed:
BLU: Unwhistled Revenge Blast suggestion was never triggering

Just a silly bug; I was trying to get the information out of the
HistoryEntry object rather than the underlaying object with the
data.

Hugmeir committed:
core Action interface: allow specifying the damage type (magical, physical, dark)

Hugmeir committed:
BLU: mark damage-dealing actions with their damage type (physical/magical/dark)

Hugmeir committed:
BLU: Raid buffs, sort of

Due to the way Off-guard and Peculiar Light work, it is possible
to have 100% uptime of both during a fight.

So a plain "you popped your buff on cooldown" isn't quite the ticket
here -- what this report shows is how many actions were buffed by
applying the buff, and of those, how many were also covered by
someone else applying the buff.

If the numbers don't match, then you and your Buff Buddy(tm) are
missing some coordination.  And if the numbers are low, then either
downtime is involved, or the group can spot some improvement points.

Hugmeir committed:
BLU: Conditionally enable the Overheal report

The report only shows up if they did any non-White Wind heals.
Ideally this would be toggled based on mimickry, but detection
of the status is pretty finicky (some pulls have it, some don't)
so this is a compromise.

Hugmeir committed:
BLU: Detect when Triple Trident is used without Whistle and Tingle

Triple Trident can be loosely thought of as a combo finisher;
you can use it on its own but you lose a ton of potency by not
using Whistle and Tingle first.

Hugmeir committed:
BLU: Track dropped/overwritten buffs

"You cast it, use it!"

Hugmeir committed:
ExpectedActionsEvaluator: allow overriding the icon shown for each action

This is actually groundwork for some BLU tweaks; BLU has mutually
exclusive abilities (e.g. J Kick / Quasar), and we need to check
if either is used, rather than both.

Hugmeir committed:
BLU: Extend the Moon Flute report to support alternative actions.

Basically, all of these are mutually exclusive abilities that
share the same cooldown:

    J Kick       / Quasar
    Eruption     / Feather Rain
    Shock Strike / Mountain Buster

Using either during a moon flute is acceptable, so we should track
both.

Note that we explicitly do not track Both Ends / Nightbloom,
because Both Ends is strictly worse in almost every raiding
situation -- When are you ever going to do a fight that
has 3+ targets during every burst window?

Hugmeir committed:
BLU Revenge Blast report: handle auto-attacks & sprint breaking the report

Hugmeir committed:
BLU: Time spent doing the x4 Surpnakha combo is uptime

Hugmeir committed:
BLU Raid buffs: Take into account the Condensed Libra buffs

Hugmeir committed:
Raid Buffs: Display the Condensed Libra buffs in the timeline

Hugmeir committed:
Add myself to CONTRIBUTORS and as a DEVELOPER for the Blue Mage page

Yes, my icon is a red Flying Sardine.  I'm really just a Red Mage
moonlighting as a Blue Mage, after all.

Hugmeir committed:
BLU: Mark 6.0 - 6.3 as supported patches

Hugmeir committed:
BLU raid buffs: was using an undefined value was a row key

Hugmeir committed:
BLU DoTs: remove a leftover this.debug()

Hugmeir committed:
BLU: Add all the new Trans entries to locale/en/messages.json

Hugmeir committed:
BLU: Note in the changelog that BLU had some changes

Hugmeir committed:
BLU: Minor `yarn lint --fix` fixes

Hugmeir committed:
BLU: tsc linter issues in the Weaving module

Hugmeir committed:
BLU: tsc linter issues in the RevengeBlast module

Hugmeir committed:
BLU: tsc linter issues in the ABC module

Hugmeir committed:
BLU: tsc linter issues in the Overheal module

Hugmeir committed:
BLU: tsc linter issues in the MoonFlute module

Hugmeir committed:
BLU: tsc linter issues in the DroppedBuffs module

Hugmeir committed:
BLU: tsc linter issues in the ColdFog module

Hugmeir committed:
BLU: Fixed typo in one of the Song of Torment suggestions

Hugmeir committed:
BLU: tsc linter issues in the DoT module

Hugmeir committed:
BLU: tsc linter issues in the BLURaidBuffs module

Hugmeir committed:
BLU changelog: Fix a typo

Hugmeir committed:
BLU: Use the Swiftcast module in a way that typescript won't barf at

Hugmeir committed:
BLU: Minor grammar fixup in the About section

Hugmeir committed:
BLU: yarn extract

Hugmeir committed:
BLU: Bump support for 6.4

Hugmeir committed:
BLU: Fix a typo in one translation tag

Hugmeir committed:
BLU: Move our defensives to the new Defensives module

Hugmeir committed:
BLU Moon Flute: Tell ExpectedGcdCountOptions that Moon Flute does not have stacks

lailleo commented on src/data/ACTIONS/root/BLU.ts:
```diff
@@ -604,4 +607,211 @@ export const BLU = ensureActions({
+		icon: 'https://xivapi.com//i/003000/003355.png',
+		gcdRecast: 2500,
+		cooldown: 120000,
+		onGcd: true, // not actually true but makes the timeline nicer
```
Agreed that this likely will make the report more readable instead of looking like a huge gap.

lailleo commented on src/parser/jobs/blu/index.tsx:
```diff
@@ -6,10 +6,11 @@ import React from 'react'
 import {changelog} from './changelog'
 
-const description = t('blu.about.description')`This is a limited analyzer for a limited job. It's not ready yet, but please look forward to it.
+const description = t('blu.about.description')`BLU has wildly different play styles depending on your mimickry.  This page will primarily focus on the DPS aspect of the job, so do take the advice with a grain of salt if you are healing or tanking.
```
Minor thing, but mimicry.

lailleo commented on src/parser/jobs/blu/index.tsx:
```diff
@@ -6,10 +6,11 @@ import React from 'react'
 
-This analyzer will focus on DoTs, uptime, and proper usage of [~action/BRISTLE],[~action/MOON_FLUTE], and [~action/SURPANAKHA].
-`
+Worth also keeping in mind the usual adage that BLU content is won and lost on mechanics, not DPS.  Following the suggestions here to optimize your performance will always be welcome, but it should not detract from mechanics; greeding Phantom Flurry ticks at the expense of your team mate's sanity is not recommended.
```
Another tiny little thing, teammates'

Hugmeir replied:
Patched in ff3f05c59604bb1ac24a4e8aa90d275f2920cf0c

lailleo commented on locale/en/messages.json:
```diff
@@ -165,8 +165,55 @@
   "blm.triplecast.suggestions.overwrote-triplecasts.why": "You overwrote <0/> .",
-  "blu.about.description": "This is a limited analyzer for a limited job. It's not ready yet, but please look forward to it.\n\nThis analyzer will focus on DoTs, uptime, and proper usage of [~action/BRISTLE],[~action/MOON_FLUTE], and [~action/SURPANAKHA].\n",
-  "blu.about.description.warning": "This module is under development and not yet ready for the public.",
+  "blu.about.description": "BLU has wildly different play styles depending on your mimickry.\n\nThis page will primarily focus on the DPS aspect of the job, so do take the advice with a grain of salt if you are healing or tanking.\n\nWorth also keeping in mind the usual adage that BLU content is won and lost on mechanics, not DPS.  Following the suggestions here to optimize your performance will  ...
```
Mimicry, mentioned in another comment as well.

Hugmeir replied:
Patched in ff3f05c59604bb1ac24a4e8aa90d275f2920cf0c

AkairyuGestalter commented on src/parser/jobs/blu/modules/MoonFlute.tsx:
```diff
@@ -0,0 +1,179 @@
+			hasStacks: false,
+		}))
+
+		const mfActionEvaluator = new MoonFluteExpectedActionsEvaluator({
```
Rather than rolling your own extension of the ExpectedActionsEvaluator, you probably ought to use the ExpectedActionGroupsEvaluator. That should get you what you're looking for.

Hugmeir replied:
Patched in 818143d07c71f7336e42a0f763da787778ffeed4

AkairyuGestalter commented on src/parser/jobs/blu/modules/Overheal.tsx:
```diff
@@ -0,0 +1,77 @@
+
+export class BLUOverheal extends Overheal {
+
+	@dependency private actors!: Actors
```
Unused dependency, can get rid of this.

Hugmeir replied:
Patched in 6c55c596f9cf850dc7756243e0fe541b939e343f

AkairyuGestalter commented on src/parser/jobs/blu/modules/AlwaysBeCasting.tsx:
```diff
@@ -0,0 +1,195 @@
+				Dropping out of <DataLink action="PHANTOM_FLURRY" /> too early will lose damage ticks.  If you are in a <DataLink action="MOON_FLUTE" /> window you want to wait out the entire channel; if you are using it outside of a window and activating the final kick, wait until the last second the <DataLink status="PHANTOM_FLURRY" /> effect is active.
+			</Trans>,
+			why: <Trans id="blu.phantom_flurry.dropped_ticks.why">
+				<Plural value={missingFlurryTicks ?? 0} one="# Phantom Flurry tick was" other="# Phantom Flurry ticks were" /> dropped due to cancelling the channel too early.
```
Not sure what was here before, but you don't need a null-coalesce since the assignment will always give you a number (even if that number is 0).

Hugmeir replied:
Patched in 430831ee2ee7ad6bc6def291be7bc54f92ddde46

AkairyuGestalter commented on src/parser/jobs/blu/modules/AlwaysBeCasting.tsx:
```diff
@@ -0,0 +1,195 @@
+				While the channel from <DataLink action="PHANTOM_FLURRY" /> is active, it becomes <DataLink action="PHANTOM_FLURRY_KICK" />, a 600 potency button.  If you are using <DataLink action="PHANTOM_FLURRY" /> outside of a <DataLink action="MOON_FLUTE" showIcon={false} /> window, then you should use the 600 potency button before the channel runs out.  Use this even if it means dropping the last tick  ...
+			</Trans>,
+			why: <Trans id="blu.phantom_flurry.dropped_kicks.why">
+				<Plural value={missingFlurryKicks ?? 0} one="# Phantom Flurry big kick was" other="# Phantom Flurry big kicks were" /> dropped by not pressing the button again before the effect ran out.
```
Same note here about not needing the null-coalesce.

AkairyuGestalter commented on src/parser/jobs/blu/modules/AlwaysBeCasting.tsx:
```diff
@@ -0,0 +1,195 @@
+		this.suggestions.add(new TieredSuggestion({
+			icon: this.data.actions.PHANTOM_FLURRY.icon,
+			content: <Trans id="blu.phantom_flurry.dropped_ticks.content">
+				Dropping out of <DataLink action="PHANTOM_FLURRY" /> too early will lose damage ticks.  If you are in a <DataLink action="MOON_FLUTE" /> window you want to wait out the entire channel; if you are using it outside of a window and activating the final kick, wait until the last second the <DataLink status="PHANTOM_FLURRY" /> effect is active.
```
Basically all of the rest of Xiva's copy does not use double-spaces after a period. Can we go to single-spaces for consistency?

AkairyuGestalter commented on src/parser/jobs/blu/modules/BLURaidBuffs.tsx:
```diff
@@ -0,0 +1,442 @@
+				<Trans id="blu.buffs.table.message">
+				Blue Mages can keep both <ActionLink action="OFF_GUARD"/> and <ActionLink action="PECULIAR_LIGHT"/> up for the entire duration of the fight, and may opt to have the <ActionLink action="CONDENSED_LIBRA"/> buffs running as well.
+					<br />
+				The table below shows when you used your buffs, as well as how many damaging party actions the buff covered; it also shows how many of those actions were also covered by the other two buffs.  Ideally your team should coordinate to have <ActionLink action="OFF_GUARD" showIcon={false} /> and <ActionLink action="PECULIAR_LIGHT" showIcon={false} /> running all the time, making both numbers below  ...
```
Just use periods instead of semi-colons. The simpler the copy can be, the easier it'll be for users to read and for translators to do their thing.

Hugmeir replied:
Patched in 50774b28a0a00c0ce11701bf8d69fe68455d1daf

AkairyuGestalter commented on src/parser/jobs/blu/modules/ColdFog.tsx:
```diff
@@ -0,0 +1,146 @@
+	static override handle = 'coldfog'
+	static override title = t('blu.cold_fog.title')`Cold Fog`
+
+	@dependency private actors!: Actors
```
Unused dependency, can drop this.

Hugmeir replied:
Patched in fe8b2e5eeca640fc183bb44db11cc616dd174ba5

AkairyuGestalter commented on src/parser/jobs/blu/modules/ColdFog.tsx:
```diff
@@ -0,0 +1,146 @@
+	@dependency private data!: Data
+	@dependency private suggestions!: Suggestions
+
+	private touchOfFrostProcs = 0
```
Looks like this variable isn't used anymore either?

Hugmeir replied:
Patched in 4c257c4f7d9a3c05b398253c8f79da1d9557397d

AkairyuGestalter commented on src/parser/jobs/blu/modules/ColdFog.tsx:
```diff
@@ -0,0 +1,146 @@
+			icon: this.data.actions.COLD_FOG.icon,
+			content: <Trans id="blu.coldfog.dropped.content">
+				Any damage taken while the <DataLink status="COLD_FOG"/> effect is active grants the <DataLink status="TOUCH_OF_FROST"/>
+				buff, allowing the use of <DataLink action="WHITE_DEATH" /> for 15 seconds.
```
If you want the newlines to actually show in the end result output, don't forget to include the </br> elements at the end of each line. As it is, the next line is just going to directly append itself to the previous and you'll lose some of the intended whitespace.

Hugmeir replied:
Yup, had a missing `br`, patched in f1fc88138da52e821897195b74827ce058a041ce

AkairyuGestalter commented on src/parser/jobs/blu/modules/MoonFlute.tsx:
```diff
@@ -0,0 +1,179 @@
+			globalCooldown: this.globalCooldown,
+			suggestionIcon,
+			suggestionContent: <Trans id="blu.moonflutes.suggestions.gcds.content">
+				Regardless of spell speed, ideally a <ActionLink action="MOON_FLUTE" /> window should contain at least
```
This (and other suggestions) are *very* heavy on links with icons in them, which can get pretty noisy to look at. You might consider adding a showIcon={false} property to some of these to cut down on the visual clutter (especially if you have the same action or status linked multiple times in the same suggestion).

Hugmeir replied:
I've sprinkled some showIcon false all over the place in 7625051a77f82f8daa110e5c408209fc9e88619c

Mentioned this on the discord, but the one hill I'm dying on is the ColdFog suggestions. Had several cases of folk who knew to proc Cold Fog, but didn't realize it gave you a new buff that allowed you to cast a different GCD, so I want the icons big and under the spotlight.

AkairyuGestalter commented on src/parser/jobs/blu/modules/RevengeBlast.tsx:
```diff
@@ -0,0 +1,221 @@
+				// TODO: this is a cop-out.  Dealing with downtime is hard.
+				if (forcedDowntime > REVENGE_BLAST_MINIMUM_WINDOW_MS) { return }
+
+				const deltaMs      = revengeEnd - revengeStart - forcedDowntime
```
I'm mildly surprised the linter isn't complaining about this, but adding all the extra whitespace between the `const varName` and the `= <assignment>` chunks makes this harder to read. Can we get rid of that (and double-check the other files as well, since I saw the same pattern elsewhere)?

Hugmeir replied:
Oh that's my work linter for another language firing by mistake; it was trying to whitespace-align operators, ala
```
const foo     = 1
const barquux = 2
```

Anyway, should be removed as of 58930091f8629afd9498f63502065dffba1a8507

AkairyuGestalter commented on src/parser/jobs/blu/modules/Swiftcast.tsx:
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+
+export class Swiftcast extends CoreSwiftcast {
+	override suggestionContent = <Trans id="blu.swiftcast.suggestion.content">Cast a spell with <DataLink action="SWIFTCAST" /> before it expires.</Trans>
```
Do we need this suggestion override? The core one gives the additional context about using it for movement or weaving that this one lacks.

Hugmeir replied:
With the exception of the RDM module, this is actually something cargo-culted by every other job!

The underlaying issue is that simply using the core module directly makes `tsc` unhappy[0], so you need to override the class; everyone likely just copypasted the same class override that also happens to override the core suggestion.

Should be patched as of 45b06261776e1e69de63ac4eda032075042b6abe 

[0]
```
; tsc
src/parser/jobs/blu/index.tsx:16:17 - error TS2322: Type 'Promise<typeof import("./src/parser/jobs/blu/modules/index")>' is not assignable to type 'Promise<{ default: (typeof Injectable)[]; }>'.
  Type 'typeof import("./src/parser/jobs/blu/modules/index")' is not assignable to type '{ default: (typeof Injectable)[]; }'.
    Types of property 'default' are incompatible.
      Type '(typeof Tincture | typeof Swiftcast | typeof AlwaysBeCasting | typeof BLURaidBuffs | typeof ColdFog | typeof Defensives | typeof DoTs | typeof DroppedBuffs | typeof GeneralCDDowntime | typeof Interrupts | typeof BLUOverheal | typeof RevengeBlast | typeof TripleTrident | typeof BLUWeaving)[]' is not assignable to type '(typeof Injectable)[]'.
        Type 'typeof Tincture | typeof Swiftcast | typeof AlwaysBeCasting | typeof BLURaidBuffs | typeof ColdFog | typeof Defensives | typeof DoTs | typeof DroppedBuffs | typeof GeneralCDDowntime | typeof Interrupts | typeof BLUOverheal | typeof RevengeBlast | typeof TripleTrident | typeof BLUWeaving' is not assignable to type 'typeof Injectable'.
          Type 'typeof Swiftcast' is not assignable to type 'typeof Injectable'.
            Cannot assign an abstract constructor type to a non-abstract constructor type.

16  modules: () => import('./modules' /* webpackChunkName: "jobs-blu" */),
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  src/parser/core/Meta.tsx:9:22
    9 type ModulesLoader = () => Promise<{default: Array<typeof Injectable>}>
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The expected type comes from the return type of this signature.


Found 1 error.

```

AkairyuGestalter commented on scripts/generate-unable-to-act.ts:
```diff
@@ -64,6 +64,18 @@ async function main() {
 		})
 	}
 
+    // The secondary effect from Moon Flute, Waning Nocturne, makes you
```
VS Code's linter is complaining about this indentation being 4 spaces instead of a tab. Not sure why github passed it but we should probably fix it anyways.

Hugmeir replied:
Ah, `yarn lint` only applies to `src/` apparently, so `scripts/` escaped notice.
Patched with a503c399d8a59fb52a530691fea6df9662acc4f7 and 5ee999fb

Probably worth patching package.json to include `scripts/`

AkairyuGestalter reviewed (changes requested):
Couple comments, mostly around the content and formatting of the copy, and some other code style things. I don't have any complaints about the functionality itself. I only left one instance of most of the comments about the copy formatting, but a lot of them apply somewhat across the board so do give all of the files a once over with each of them in mind!

Hugmeir committed:
BLU: single spaces after each period

Hugmeir committed:
BLU: mimickry => mimicry, also teammate's => teammates'

Hugmeir committed:
BLU: so long, copy semicolons; we did not deserve you.

Hugmeir committed:
BLU Overheal report: Remove unneeded dep on Actors

Hugmeir committed:
BLU ColdFog module: Remove unneeded dep on Actors

Hugmeir committed:
BLU ColdFog: missing newline in copy

Hugmeir committed:
scripts/generate-unable-to-act.ts: spaces => tabs

Hugmeir committed:
BLU: Remove operator-alignment whitespace

Hugmeir committed:
BLU: Remove the suggestion override from the Swiftcast subclass

Sadly can't remove the entire subclass entirely, tsc does not like
us using the core module directly.

Hugmeir committed:
BLU: no need to null-coalesce Plural value variables, holdover from an old mistake

Hugmeir committed:
BLU Cold Fog: remove unused counter attribute

Hugmeir committed:
BLU: Sprinkling showIcon=false over a bunch of suggestions

Hugmeir committed:
BLU Moon Flute: use ExpectedActionGroupsEvaluator instead of rolling our own subclass

Hugmeir committed:
BLU Cold Fog: minor copy whitespace tweak

Hugmeir committed:
BLU: update all messages.json files via yarn extract

Hugmeir committed:
BLU: yarn lint fixes

Hugmeir committed:
scripts/generate-unable-to-act.ts: addressing an other minor linter issue

AkairyuGestalter reviewed (approved):
Happy with the cleanup! Anything else can be polished on subsequent PRs 😃
